### PR TITLE
Fix disconnected graph bug in link prediction CCPA and add tests

### PR DESCRIPTION
**Pull request by Zmeos**
(Note: the question at the bottom should be discussed and addressed before merging)

Fixes a bug where CCPA throws a KeyError due to there being no path between nodes. I assume that the method should return a score of 0 if there is no path between two nodes. I make this assumption based on this being a local neighbourhood similarity score based on common neighbours and shortest path. CCPA gives a higher score for nodes that have a small shortest path and a lower score for a longer shortest path. No shortest path means the nodes have no structural connection and also no common neighbours, thus a score of 0 is sensible. This is in line with other link prediction methods (based on common neighbours) which simply returns a score of 0 inherently, due to there being no common neighbours.

I suspect that this would normally have been caught if the method had tests. So I added tests. The bug fixed is covered by the "test_no_common_neighbor" test.

The behaviour can be replicated with the following code.
```python
import networkx as nx
G = nx.Graph()
G.add_nodes_from([0, 1])
list(nx.common_neighbor_centrality(G))
```
Expected behaviour, returns: `[(0, 1, 0)]` 
Observed behaviour, throws KeyError.

Before this is merged, I have a question. Other link prediction methods were tested on links between equal nodes (the `test_equal_nodes` test). CCPA in its current state throws a ZeroDivisionError on this test, since the shortest path is zero. I've added the test but commented it out. I'll adapt the test based on what we decide. 
It is less clear to me what a sensible score for equal nodes should be as I'm not familiar with any use case for such a prediction. I see at least two options:
1. Leave the functionality as is and modify the test to expect an exception. In this case, should the method call fail with an informative error and error message, which one?
2. Assume a score for this case. In this case, what would a sensible score be?

### CODE REVIEW

**Comment by rossbar** [comment]
Thanks for bringing this up. I'm not too familiar with this particular centrality measure so I don't have an immediate sense for what makes sense: should the CCPA result be `0` in this case or is it rather undefined? Does the Ahmad, Akhtar, Noor et. al. paper mention anything about this case (i.e. unconnected nodes)?

**Comment by Zmeos** [comment]
Thank you for answering. I've had another look at the paper ([https://www.nature.com/articles/s41598-019-57304-y](link) for convenience). I cannot see them mentioning either corner case. I'll try to outline my thoughts a bit more succinctly on why I think this is the right choice, I'm happy to discuss more and possibly leave both corner cases undefined (although we should at least keep some tests).

**Corner case 1. Unconnected nodes**
- CCPA weights the contribution the common neighbours similarity measure and the length of the shortest path to generate a similarity score for two nodes. If there is no path between those nodes the common neighbours measure will inherently report a score of 0 (there are no neighbours) and the shortest path does not exist. In terms of similarity I find it fair to assume that the similarity should then be 0. This is also the lower bound of the similarity measure.
- Here is another intuitive way of thinking about it. When two nodes are close, the common neighbour measure dominates the scoring, but when nodes do not have common neighbours, the distance between the nodes is used to score the similarity. The further away, the lower the similarity. If there is no path, there is no similarity. No similarity is a similarity of 0.

**Corner case 2: Same nodes**
I find it trickier to come up with a good score in this case. After some consideration I think the best choice in this case is to leave it undefined as this will only occur if the user explicitly requests these links with the ebunch parameter.

**Comment by dschult** [comment]
See my [comment](https://github.com/networkx/networkx/issues/4956#issuecomment-888668233) in #4956 where I suggest that the proper "distance" between nodes from different components is infinity.  Then the portion of the score due to distance is zero (and common neighbors is also zero because they can't have common neighbors either). 

I think the right fix is to change the shortest_path dictionary to hold path lengths and then use `inf = float('inf')` and
```python
    spl = nx.shortest_path_length(G)

    def predict(u, v):
        return alpha * sum(1 for _ in nx.common_neighbors(G, u, v)) + (1 - alpha) * (len(G) / (spl[u].get(v, inf) - 1)
```

**Comment by Zmeos** [comment]
Thank you for noticing the related issue and your reply. I agree that the distance contribution to the similarity score should be zero. 

I've now implemented your suggested solution for the disconnected components case. For the self link case (division by zero issue) I've simply added an algorithm exception, I'd be happy to discuss this if anyone have a good suggestion for which score should be returned in that case. I've also added a check for valid alphas and a test for that check.

**Comment by dschult** [comment]
Did you mean to change the denominator to path_len instead of path_len - 1?
Also, you should define float(“inf”) outside the function call so it only has to be defined once (the second argument to get is computed before calling… one of the downsides of the get method).

**Comment by Zmeos** [comment]
The denominator change was intentional. Since the shortest_path_length returns the length of the shortest path there was no longer the need to subtract by 1 to get the correct length.

Good point on the float("inf"). I've updated that. I also missed that you used a generator expression to count the number of neighbours. I've included that now.

File: networkx/algorithms/link_prediction.py [hunk 1]
```diff
@@ -275,11 +275,20 @@ def common_neighbor_centrality(G, ebunch=None, alpha=0.8):
            Sci Rep 10, 364 (2020).
            https://doi.org/10.1038/s41598-019-57304-y
     """
-    shortest_path = nx.shortest_path(G)
+
+    if not 0 <= alpha <= 1:
+        raise nx.NetworkXAlgorithmError("Alpha must be between 0 and 1")
```
Comment (dschult) [inline networkx/algorithms/link_prediction.py:280]: Do we need to restrict the values of alpha?
I know that traditionally the values would be between 0 and 1. But I've often seen algorithms with a weighted average like this be "over-compensated" by choosing weights that are either negative or more than one. This can sometimes successfully measure a feature.  It's like putting a negative weight on common_neighbors with compensating positive weight on closeness.  

If we're not sure, I'd prefer to add a doc_string line: "Values for alpha should normally be between 0 and 1."
and take out the explicit check and raise exception code.

Comment (Zmeos) [reply networkx/algorithms/link_prediction.py]: Good point, I had not considered that. I've removed the restriction code and the test for it. I've also added the sentence suggesting normal values to the doc string.

**Review by dschult — APPROVED** [review]
This looks ready to merge to me.
Thanks!

File: networkx/algorithms/link_prediction.py [hunk 2]
```diff
@@ -275,11 +276,17 @@ def common_neighbor_centrality(G, ebunch=None, alpha=0.8):
...
-    shortest_path = nx.shortest_path(G)
+
+    spl = dict(nx.shortest_path_length(G))
+    inf = float("inf")
 
     def predict(u, v):
-        return alpha * len(list(nx.common_neighbors(G, u, v))) + (1 - alpha) * (
-            G.number_of_nodes() / (len(shortest_path[u][v]) - 1)
+        if u == v:
+            raise nx.NetworkXAlgorithmError("Self links are not supported")
```
Comment (rossbar) [inline networkx/algorithms/link_prediction.py:285]: Do any of the measures in `link_prediction` support self-loops? If not, it might be nice to see if this check could be moved out to `_apply_prediction`. Let's worry about that in a follow-up PR though.

File: networkx/algorithms/tests/test_link_prediction.py [hunk 3]
```diff
@@ -165,6 +165,57 @@ def test_all_nonexistent_edges(self):
...
+            self.func,
+            nx.MultiGraph([(0, 1), (1, 2)]),
+            [(0, 2)],
+        )
+        assert pytest.raises(
+            nx.NetworkXNotImplemented,
+            self.func,
+            nx.MultiDiGraph([(0, 1), (1, 2)]),
+            [(0, 2)],
+        )
```
Comment (rossbar) [inline networkx/algorithms/tests/test_link_prediction.py:202]:
```suggestion
    @pytest.parametrize(
        'graph_type', (nx.DiGraph, nx.MultiGraph, nx.MultiDiGraph)
    )
    def test_notimplemented(self, graph_type):
        assert pytest.raises(
            nx.NetworkXNotImplemented, self.func, graph_type([(0, 1), (1, 2)]), [(0, 2)]
        )
```

I realize this test was just copied from the others in the file, but might as well save a little repetition by parametrizing :)

**Review by rossbar — APPROVED** [review]
LGTM - great idea about the infinite distance between disconnected components. Thanks @Zmeos & @dschult !

File: networkx/algorithms/link_prediction.py (hunk 2 above)
Comment (Zmeos) [reply networkx/algorithms/link_prediction.py:285]: I just had a quick look. It looks like all the other measures do support self loops. This is tested by the "test_equal_nodes" tests. Interestingly the Jaccard coefficient and Preferential attachment methods also support self loops although they don't have the "test_equal_nodes" test. I might add a follow up PR to clean up the link prediction tests.

File: networkx/algorithms/tests/test_link_prediction.py (hunk 3 above)
Comment (Zmeos) [reply networkx/algorithms/tests/test_link_prediction.py]: Great catch! I'm tempted to add this change to all the link prediction tests now.

**Comment by rossbar** [comment]
Oops - my suggestion for parametrization was wrong... sorry about that. I've pushed up the fix and we can (finally) push the green button. Thanks @Zmeos !
